File: hydrowin/hydrowin/backend/app/live_telemetry.py
"""Кэш последних показаний на sensors + сборка live-ответов API."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from sqlalchemy.orm import Session

from app.deps import iso_z
from app.models import Machine, Sensor, User
from app.org_access import machine_access_for_user, machines_visible_query
from app.security import effective_machine_status


def apply_sensor_live(
    sensor: Sensor,
    *,
    value: float,
    ts: datetime,
    status: str,
    fault: str | None = None,
    current_ma: float | None = None,
) -> None:
    sensor.last_value = value
    sensor.last_ts = ts
    sensor.last_status = status
    sensor.last_fault = fault
    sensor.last_current_ma = current_ma


def sensor_live_json(sensor: Sensor) -> dict[str, Any]:
    return {
        "id": sensor.id,
        "channel": sensor.channel_index,
        "name": sensor.name,
        "type": sensor.type,
        "unit": sensor.unit,
        "value": sensor.last_value,
        "ts": iso_z(sensor.last_ts) if sensor.last_ts else None,
        "status": sensor.last_status or "offline",
        "fault": sensor.last_fault,
        "current_ma": sensor.last_current_ma,
    }


def machine_live_json(machine: Machine, sensors: list[Sensor]) -> dict[str, Any]:
    ordered = sorted(sensors, key=lambda s: s.channel_index)
    return {
        "machine_id": machine.id,
        "code": machine.code,
        "status": effective_machine_status(machine),
        "last_seen_at": (
            iso_z(machine.last_seen_at) if machine.last_seen_at else None
        ),
        "sensors": [sensor_live_json(s) for s in ordered],
    }


def fleet_live_for_user(db: Session, user: User) -> dict[str, Any]:
    machines = machines_visible_query(db, user).order_by(Machine.code).all()
    machine_ids = [m.id for m in machines]
    sensors_by_machine: dict[str, list[Sensor]] = {mid: [] for mid in machine_ids}
    if machine_ids:
        for sensor in (
            db.query(Sensor).filter(Sensor.machine_id.in_(machine_ids)).all()
        ):
            sensors_by_machine.setdefault(sensor.machine_id, []).append(sensor)

    items: list[dict[str, Any]] = []
    for machine in machines:
        if machine_access_for_user(db, user, machine) is None:
            continue
        items.append(
            machine_live_json(machine, sensors_by_machine.get(machine.id, []))
        )

    return {
        "machines": items,
        "at": iso_z(datetime.utcnow()),
    }

File: hydrowin/hydrowin/mobile/hydrowin/lib/domain/models/live_telemetry.dart
import 'package:hydrowin/domain/models/machine_summary.dart';

class SensorLiveReading {
  const SensorLiveReading({
    required this.id,
    required this.channel,
    required this.name,
    required this.unit,
    this.value,
    this.ts,
    this.status = 'offline',
    this.fault,
    this.currentMa,
  });

  final String id;
  final int channel;
  final String name;
  final String unit;
  final double? value;
  final DateTime? ts;
  final String status;
  final String? fault;
  final double? currentMa;

  factory SensorLiveReading.fromJson(Map<String, dynamic> json) {
    final rawTs = json['ts'] as String?;
    return SensorLiveReading(
      id: json['id'] as String,
      channel: json['channel'] as int? ?? 0,
      name: json['name'] as String? ?? '',
      unit: json['unit'] as String? ?? '',
      value: (json['value'] as num?)?.toDouble(),
      ts: rawTs != null ? DateTime.tryParse(rawTs) : null,
      status: json['status'] as String? ?? 'offline',
      fault: json['fault'] as String?,
      currentMa: (json['current_ma'] as num?)?.toDouble(),
    );
  }
}

class MachineLiveSnapshot {
  const MachineLiveSnapshot({
    required this.machineId,
    required this.code,
    required this.status,
    required this.sensors,
    this.lastSeenAt,
  });

  final String machineId;
  final String code;
  final MachineStatus status;
  final DateTime? lastSeenAt;
  final List<SensorLiveReading> sensors;

  factory MachineLiveSnapshot.fromJson(Map<String, dynamic> json) {
    final sensorsJson = json['sensors'] as List<dynamic>? ?? const [];
    return MachineLiveSnapshot(
      machineId: json['machine_id'] as String,
      code: json['code'] as String? ?? '',
      status: machineStatusFromString(json['status'] as String? ?? 'offline'),
      lastSeenAt: json['last_seen_at'] != null
          ? DateTime.tryParse(json['last_seen_at'] as String)
          : null,
      sensors: sensorsJson
          .map((e) => SensorLiveReading.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

class FleetLiveSnapshot {
  const FleetLiveSnapshot({
    required this.machines,
    this.at,
  });

  final List<MachineLiveSnapshot> machines;
  final DateTime? at;

  factory FleetLiveSnapshot.fromJson(Map<String, dynamic> json) {
    final machinesJson = json['machines'] as List<dynamic>? ?? const [];
    return FleetLiveSnapshot(
      machines: machinesJson
          .map((e) => MachineLiveSnapshot.fromJson(e as Map<String, dynamic>))
          .toList(),
      at: json['at'] != null
          ? DateTime.tryParse(json['at'] as String)
          : null,
    );
  }
}
